### ISSUE TITLE
Demo error: [Error] Failed to load resource: Origin http://vue.layer7.be is not allowed by Access-Control-Allow-Origin. (login, line 0)

### ISSUE THREAD

**Issue by iannazzi** [layer7be/vue-starter#18]
I am struggling to get the vue app talking with the laravel app on my local machine using a variety of virtual machines including docker containers, local machine, and homestead. 

I noticed an error on the demo site when posting login credentials. I have the same issue as the demo site
`[Error] Failed to load resource: Origin http://vue.layer7.be is not allowed by Access-Control-Allow-Origin. (login, line 0)`

And the error I am getting when attempting to post login credentials:
`[Error] Failed to load resource: Origin http://192.168.10.10:8888 is not allowed by Access-Control-Allow-Origin. (login, line 0)`

**Comment by silverark** [comment]
You could try the below in your htaccess file
'''
<IfModule mod_headers.c>
    Header set Access-Control-Allow-Origin "*"
</IfModule>
'''

**Comment by lukepomeroy** [comment]
I have also struggled with a similar error message relating to cors headers...

I used Postman to query the api as I was getting a 500 error in Chrome and then got a 'class cors could not be found' error.

Found that the issue is with kernel routes for cors, see:
[https://github.com/barryvdh/laravel-cors/issues/109]

Perhaps your issue is different, but worth trying the temp fix suggested in the link.

**Comment by kcalliauw** [comment]
merged a fixed for this and the JWTAuth middleware location. Please try again and let me know if it worked. Thanks!

**Comment by iannazzi** [comment]
thanks, let me take a look in the morning. @silverark one of my virtual machines is using nginx, and I believe laravel homestead uses nginx, so the apache mid does not seem to apply? Maybe I can add a docker vm to the project. I do have laravel angular working well together, but really wanted to get vue up and running to compare, and I love how this project is totally separate front end and back end.

**Comment by silverark** [comment]
@iannazzi  The same should apply for nginx, but with slightly different syntax. 

add_header 'Access-Control-Allow-Origin' '*';
add_header 'Access-Control-Allow-Credentials' 'true';
add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';

http://serverfault.com/questions/162429/how-do-i-add-access-control-allow-origin-in-nginx/

**Comment by kcalliauw** [comment]
It appears there is still an issue somewhere, probably with JWTAuth. I will try to take a look at it later today. A Dockerfile would be most welcome!

**Comment by iannazzi** [comment]
@kcalliauw this is totally working now! Not sure where you see the JWTAuth issue, everything worked fine on my local machine. Virtual machine would probably be better... Let me take a few weeks and work on it.

**Comment by iannazzi** [comment]
@kcalliauw although login started working, I found this same error in the 'add a dog' area. 

I then  merged much of the dingo, cors, jwt api functionality from [https://github.com/jadjoubran/laravel5-angular-material-starter](https://github.com/jadjoubran/laravel5-angular-material-starter), however I am back to the original errors even at login without resolution.
Now being a bit lost I am looking again at the @silverark 's answer... which brings me back to vm's with a custom configuration which brings me to docker which I have has some success with, but not enough to replace homestead. I do use a docker container for mysql. I have limited success in serving an app. I could go down this path to get the app served with docker with custom nginx or apache config files.

The angular-material-starter is all in one repository. Although I like the demonstration of the separation of the front end and the back end, would it be better to merge the two using gulp to generate files and one server? Was there a more compelling reason you split them out?

**Comment by hippomundo** [comment]
@iannazzi I will look further into the issue tonight or tomorrow, it has been quite busy here lately, but I'll do my best.
The main reasoning behind splitting it into 2 projects, was that people could basically use their own backend API and completely replace the Laravel one, provided the API stayed the same. At the time, Vue was starting to get some more traction, for the sake of not pushing people in a specific direction and having them be overwhelmed by a framework they don't know fully, I though it best to split up the code so it's quite obvious what code is for the frontend, along with the workflow that goes with building it, and what code is for the backend.

**Comment by kcalliauw** [comment]
@iannazzi I will look further into the issue tonight or tomorrow, it has been quite busy here lately, but I'll do my best.
The main reasoning behind splitting it into 2 projects, was that people could basically use their own backend API and completely replace the Laravel one, provided the API stayed the same. At the time, Vue was starting to get some more traction, for the sake of not pushing people in a specific direction and having them be overwhelmed by a framework they don't know fully, I though it best to split up the code so it's quite obvious what code is for the frontend, along with the workflow that goes with building it, and what code is for the backend.

**Comment by whyspark** [comment]
@kcalliauw Looks like the classes GetUserFromToken and RefreshToken in Kernel route middleware no longer exist in the develop branch of jwt-auth, which is the branch being required in composer.json.

**Comment by kcalliauw** [comment]
Yep I noticed that as well, I reverted the dependency back to the latest in 0.5 branch or something, but there are still some issues then. Will get back on it asap

**Comment by iannazzi** [comment]
I have the docker VM working for the angular-laravel-material-starter application. I am going to attempt to modify docker nginx with @silverark recommendation to see if this can resolve the issue. Probably tmrw or this weekend.

Still a bit nervous about the separation in regards to moving both repositories to production. I may combine

**Comment by iannazzi** [comment]
@silverark Great news I have the docker vm up and running and able to modify the nginx config as you mentioned above. Following through the link and reading up I tried many variations without luck
@kcalliauw Great news I have a nice docker vm working as the server. It basically is [laradock](https://github.com/LaraDock/laradock) with a few mods. bad news I am still stuck with cors. I'll read up more. Not sure about the docker integration here, do you want me to add it in and create a pull request?

For me I guess the next step is a single repository with a gulp task, but i really like how you separated these projects out.....

**Comment by iannazzi** [comment]
Wait...I am confused.... is the client (vue) or the server (laravel) throwing the error?
```
XMLHttpRequest cannot load http://localhost:8097/api/login. Origin http://192.168.99.100:81 is not allowed by Access-Control-Allow-Origin.
``
' http://192.168.99.100:81is the server for the client (vue)

**Comment by whyspark** [comment]
@iannazzi the error is related to laravel. Some of the jwt-auth classes in the dev version of the package have changed. If you query the laravel api directly you will see the actual error being thrown, and this causes the error being thrown in the  browser . I tried reverting back to an earlier branch of jwt-auth as kcalliawu did, but hit errors on some of the other jwt-auth class names, so didn't get very far.

**Comment by silverark** [comment]
@whyspark  I was under the impressions that it is your browser that generates the CORS messages, not laravel. The CORS header should be attached and returned even if Laravel throws an error (404,401 or 500). 
@iannazzi Use an IP address, or add a hostname in your "hosts" file as Chrome doesn't play nicely with CORS on locahost. 
https://bugs.chromium.org/p/chromium/issues/detail?id=67743
You should be adding the htaccess rules to the server (laravel API). You can also test this by disabling CORS in chrome and testing it: http://stackoverflow.com/questions/3102819/disable-same-origin-policy-in-chrome
Alternately, test in firefox to see if taht is the specific isue.

**Comment by silverark** [comment]
@iannazzi Did you read the message about Chrome + CORS? The request would never arrive as CORS is stopping it in it's tracks. Can you test with Firefox and let me know if you are seeing the same issues.

**Comment by silverark** [comment]
@iannazzi I have combined the two repo in a project I've started. Feel free to test that out:
https://github.com/silverark/arkpm

**Comment by kcalliauw** [comment]
@iannazzi @silverark @whyspark 
I have pushed a new version of the Laravel backend with some minor changes. Basically the CORS problem seems to come from the fact that we now have to activate CORS globally, otherwise the OPTIONS calls are not properly interpreted or something. Seems to work for me now.
